### ISSUE TITLE
Not working with React 16

### ISSUE THREAD

**Issue by thiagolacerda** [tinymce/tinymce#4155]
I've been using TinyMCE with React for a while (**not** through tinymce-react though) but after I updated my project to use React 16 things are not working anymore.
It fails in the postrender phase, because TinyMCE is using document.getElementById to get elements back when those are not yet attached to the DOM (this call gets null back: https://github.com/tinymce/tinymce/blob/1c05b6686632fca6d7ffb3983febbbcbd740efd5/src/ui/main/ts/Control.ts#L643).

The common way to use React with TinyMCE is to call tinymce.init when componentDidMount is called, passing the id of the Element that corresponds to the component or the DOM element returned by ReactDOM.findDOMNode as the target attribute for tinymce settings.
Probably React has changed some of the timing of rendering (or maybe it always worked by luck?), but componentDidMount does not guarantee that the mounted element is already a child of the document (although the ownerDocument attribute of the element refers to the document that this element will be added to), thus the null object being returned by getEl.

Since the target element is passed to the tinymce settings, is there a way to perform any "find" operation having that target element as the root most parent and operate only within that element?

I'm happy to help with more investigation and give more details.

Thanks

**Comment by fyrkant** [comment]
Is there a special reason you aren't using @tinymce/tinymce-react that maybe we can address?

**Comment by thiagolacerda** [comment]
@fyrkant that problem happens with tinymce-react as well (I tested), since tinymce.init is called in componentDidMount as I'm doing. So the node is not yet attached to the document :(

**Comment by fyrkant** [comment]
In what way does it fail? I haven't noticed it in my testing of @tinymce/tinymce-react. Can you create a simple reproduction?

**Comment by thiagolacerda** [comment]
It fails when querying the document for node by its id and trying to set an attribute on it. The node is created but react hasn't added to the document yet.
I could not manage to create a simple reproduction, since it is only failing in a large project that I have... let me try again and see if I can make it happen in a simpler example. But anyways, in case anyone sees this, I was able to make it work by doing:

```ts
public componentDidMount (): void {
  requestAnimationFrame(() => {
    tinymce.init(settings);
  });
}
```

**Comment by fyrkant** [comment]
Cant you add the attribute in the render method?

**Comment by thiagolacerda** [comment]
I don't own that node. Let me try to explain the flow:
When I call tinymce.init inside componentDidMount (not using the requestAnimationFrame callback), tinymce internally creates a bunch of divs that will be added as a child of the target node parent.
The divs in question are those with the mce_ prefix as the id (mce_0, mce_1 and mce_2). Later, tinymce will as the document for any of those divs through document.getElementById.
Although those divs are created and add as a child of the target's parent node, they are not effective in the DOM yet, due to react way of rendering. That is the place (in postRender) which the setAttribute will fail ->
1 ask document for the element through the id
2 element is not in the document yet
3 fails trying to set an attribute on undefined

**Comment by fyrkant** [comment]
I recently changed `tinymce-react` to use the element itself - not the id - as a target, like this: 
https://github.com/tinymce/tinymce-react/blob/master/src/components/Editor.tsx#L68

Maybe that could fix your problem?

**Comment by thiagolacerda** [comment]
That was what I first tried when I saw that problem. But ends up having the same issue.
Even though you are passing an element as target, the new elements created by tinymce will not be in the DOM yet when the document.getElementById happens.

check this example

```ts
private getTarget (ref): void {
  tinymce.init({target: ReactDOM.findDOMNode(ref), ... });
}

public render (): JSX.Element {
  return (
    <div>
      <div id="tinymce_target" ref={this.getTarget}></div>
    </div>
  );
}
```

Passing tinymce_target element as the target, tinymce will create the mce divs and add before the tinymce_target element, but still they are not in the DOM. So document.getElementById still fails :(

Thanks for your prompt responses and help, btw. I really appreciate.
